### PR TITLE
feat: segregate staging and prod workflows

### PR DESCRIPTION
**Pull request by OlegDO**

### CODE REVIEW

File: .github/workflows/build-prod.yml [hunk 1]
```diff
@@ -0,0 +1,138 @@
+name: Build prod
+
+on:
+  push:
+    branches: [ prod ]
+
+jobs:
+  all-microservices:
+    name: Get all microservices
```
Comment (MatthewPattell) [inline .github/workflows/build-prod.yml]:
```suggestion
  prepare:
    name: Prepare envs
```

File: .github/workflows/build-prod.yml [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+      branch: ${{ steps.other-params.outputs.branch }}
+      docker-tag: ${{ steps.other-params.outputs.docker-tag }}
+
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 30
+
+      - id: microservices
+        run: npx @lomray/microservices-cli microservices-list
```
Comment (MatthewPattell) [inline .github/workflows/build-prod.yml:22]:
```suggestion
        run: npx @lomray/microservices-cli microservices-list
        
     - name: Branch name & docker tag & latest release sha
        id: other-params
        shell: bash
        run: |
          branch=$(echo ${GITHUB_REF#refs/heads/})
          echo "branch=$branch" >> $GITHUB_OUTPUT
```

File: .github/workflows/build-prod.yml [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+jobs:
+  all-microservices:
+    name: Get all microservices
+    runs-on: ubuntu-latest
+    outputs:
+      microservices: ${{ steps.microservices.outputs.list }}
+      microservices-spaced: ${{ steps.microservices.outputs.list-spaced }}
+      branch: ${{ steps.other-params.outputs.branch }}
+      docker-tag: ${{ steps.other-params.outputs.docker-tag }}
```
Comment (MatthewPattell) [inline .github/workflows/build-prod.yml]:
```suggestion
```

File: .github/workflows/build-prod.yml [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        run: npx @lomray/microservices-cli test --coverage
+
+      - name: Compile microservices
+        run: npx @lomray/microservices-cli build
+
+      # note: this also update package.json version
+      - name: Publish npm packages / create github release
+        run: npx @lomray/microservices-cli semantic-release
+        env:
+          GIT_BRANCH: ${{ needs.all-microservices.outputs.branch }}
```
Comment (MatthewPattell) [inline .github/workflows/build-prod.yml]:
```suggestion
          GIT_BRANCH: ${{ needs.prepare.outputs.branch }}
```

replace all places

Comment (MatthewPattell) [inline .github/workflows/build-prod.yml]:
```suggestion
          GIT_BRANCH: ${{ needs.prepare.outputs.branch }}
```

replace all places

File: .github/workflows/build-prod.yml [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    needs: all-microservices
+    runs-on: ubuntu-latest
+
+    env:
+      ONLY: ${{ needs.all-microservices.outputs.microservices-spaced }}
+
+    steps:
+      - uses: actions/checkout@v4
+
+      - run: echo "//npm.pkg.github.com/:_authToken=${{ secrets.GITHUB_TOKEN }}" > ~/.npmrc
```
Comment (MatthewPattell) [inline .github/workflows/build-prod.yml]:
```suggestion
```

File: .github/workflows/build-prod.yml [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+    env:
+      MICROSERVICE_NAME: ${{ matrix.microservice }}
+      WORK_DIR: "microservices/${{ matrix.microservice }}"
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+
+      - run: echo "//npm.pkg.github.com/:_authToken=${{ secrets.GITHUB_TOKEN }}" > ~/.npmrc
```
Comment (MatthewPattell) [inline .github/workflows/build-prod.yml]:
```suggestion
```

File: .github/workflows/build-prod.yml [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+        run: npx @lomray/microservices-cli ts-check
+
+      - name: Run microservices tests with coverage
+        run: npx @lomray/microservices-cli test --coverage
+
+      - name: Compile microservices
+        run: npx @lomray/microservices-cli build
+
+      # note: this also update package.json version
+      - name: Publish npm packages / create github release
```
Comment (MatthewPattell) [inline .github/workflows/build-prod.yml:75]: this step should set same version for all microservices

just make sure

File: .github/workflows/build-prod.yml [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+        env:
+          WORK_DIR: '${{ env.WORK_DIR }}'
+
+      - name: Publish to Registry
+        uses: elgohr/Publish-Docker-Github-Action@v5
+        env:
+          FROM_TAG: "latest${{ needs.all-microservices.outputs.docker-tag }}"
+        with:
+          name: "Lomray-Software/microservices/${{ env.MICROSERVICE_NAME }}"
+          tags: "latest${{ needs.all-microservices.outputs.docker-tag }},${{ steps.package-version.outputs.version}}"
```
Comment (MatthewPattell) [inline .github/workflows/build-prod.yml]:
```suggestion
          tags: "latest,${{ steps.package-version.outputs.version}}"
```

File: .github/workflows/build-prod.yml [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+      - name: Detect dockerfile
+        id: docker-file
+        run: npx @lomray/microservices-cli detect-docker-file
+        env:
+          WORK_DIR: '${{ env.WORK_DIR }}'
+
+      - name: Publish to Registry
+        uses: elgohr/Publish-Docker-Github-Action@v5
+        env:
+          FROM_TAG: "latest${{ needs.all-microservices.outputs.docker-tag }}"
```
Comment (MatthewPattell) [inline .github/workflows/build-prod.yml]:
```suggestion
        env:
          FROM_TAG: latest
```

**Review by MatthewPattell — CHANGES_REQUESTED** [review]

File: .github/workflows/build-prod.yml (hunk 7 above)
Comment (OlegDO) [reply .github/workflows/build-prod.yml:75]: Locally it runs as well, but for the forked repo i got this: The local branch prod is behind the remote one, therefore a new version won't be published.
You can click on this link if you wish to view it: https://github.com/OlegDO/microservices/actions/runs/8788816061/job/24117160686

File: .github/workflows/build-prod.yml [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+          tags: ${{ steps.package-version.outputs.version}}"
+          dockerfile: ${{ steps.docker-file.outputs.path }}
+          context: ${{ env.WORK_DIR }}
+          buildoptions: "--compress --force-rm --no-cache"
+          buildargs: FROM_TAG
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+          registry: ghcr.io
+
+  sonarcube:
```
Comment (MatthewPattell) [inline .github/workflows/build-prod.yml]: remove this job

**Review by MatthewPattell — APPROVED** [review]

**Review by MatthewPattell — APPROVED** [review]

**Pull request merged**